src/java/org/apache/cassandra/index/sai/utils/RangeUnionIterator.java

```diff
@@ -0,0 +1,163 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.cassandra.index.sai.utils;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.cassandra.io.util.FileUtils;
+
+/**
+ * Range Union Iterator is used to return sorted stream of elements from multiple RangeIterator instances.
+ */
+@SuppressWarnings("resource")
+public class RangeUnionIterator extends RangeIterator
+{
+    private final List<RangeIterator> ranges;
+
+    private final List<RangeIterator> toRelease;
+    private final List<RangeIterator> candidates = new ArrayList<>();
+
+    private RangeUnionIterator(Builder.Statistics statistics, List<RangeIterator> ranges)
+    {
+        super(statistics);
+        this.ranges = ranges;
+        this.toRelease = new ArrayList<>(ranges);
+    }
+
+    public PrimaryKey computeNext()
+    {
+        candidates.clear();
+        PrimaryKey candidate = null;
+        for (RangeIterator range : ranges)
+        {
+            if (range.hasNext())
+            {
+                // Avoid repeated values but only if we have read at least one value
+                while (next != null && range.hasNext() && range.peek().compareTo(getCurrent()) == 0)
+                    range.next();
+                if (!range.hasNext())
+                    continue;
+                if (candidate == null)
+                {
+                    candidate = range.peek();
+                    candidates.add(range);
+                }
+                else
+                {
+                    int cmp = candidate.compareTo(range.peek());
+                    if (cmp == 0)
+                        candidates.add(range);
+                    else if (cmp > 0)
+                    {
+                        candidates.clear();
+                        candidate = range.peek();
+                        candidates.add(range);
+                    }
+                }
+            }
+        }
+        if (candidates.isEmpty())
+            return endOfData();
+        candidates.forEach(RangeIterator::next);
+        return candidate;
+    }
+
+    protected void performSkipTo(PrimaryKey nextKey)
+    {
+        for (RangeIterator range : ranges)
+        {
+            if (range.hasNext())
+                range.skipTo(nextKey);
+        }
+    }
+
+    public void close() throws IOException
+    {
+        // Due to lazy key fetching, we cannot close iterator immediately
+        toRelease.forEach(FileUtils::closeQuietly);
```

Comment:
No, I think this was left over from a previous iteration of this class. At one point `ranges` was repopulated from `candidates` so did not contain the original ranges. 

As an aside to this `candidates` would be better initialised in the constructor as `candidates = new ArrayList<>(ranges.size());` because we know the maximum size it can be and that is never going to be so big that we would be better letting it grow.